### ISSUE TITLE
Error on InitiateAuthCommand with good cred >> NotAuthorizedException: Incorrect username or password >> But cr

### ISSUE THREAD

**Issue by smichel-amiltone** [aws/aws-sdk-js#4642]
### Describe the bug

Login fail with correct credential who are set in cognito user

### Expected Behavior

Should succes with TOTO123456+ Azlsd45$q#s and have AuthenticationResult in result

If i set TOTO123456 + Azlsd45$ this is working fine...

### Current Behavior

Result -> NotAuthorizedException: Incorrect username or password.

### Reproduction Steps

**Package** 
```
"@aws-sdk/client-cognito-identity-provider": "^3.245.0",
```

**Step one set a password with spécial like $ or # in the middle on some cognito user**
```
aws cognito-idp admin-set-user-password \
--user-pool-id XXXXXXXXX \
--username TOTO123456\
--password Azlsd45$q#s \
--permanent \
--profil XXXXXXX  \
--region eu-central-1
```

**JS LAMBDA SAMPLE**
```
import { CognitoIdentityProviderClient, InitiateAuthCommand } from '@aws-sdk/client-cognito-identity-provider';
import { APIGatewayEvent, Context, APIGatewayProxyResult, Handler } from 'aws-lambda';
import { COGNITO_APP_CLIENT_ID } from 'src/shared/process-env-var';

export const login: Handler = async (event: APIGatewayEvent, context: Context): Promise<APIGatewayProxyResult> => {
  try {
  
    const eventBody = parseBody(event);
    DEBUG && console.debug('>> body >>', eventBody);
    const { email: USERNAME, password: PASSWORD } = eventBody;


    const client = new CognitoIdentityProviderClient({});
    const AuthParameters = {
      USERNAME,
      PASSWORD,
    };
    const input = {
      ClientId: COGNITO_APP_CLIENT_ID,
      AuthFlow: 'USER_PASSWORD_AUTH',
      AuthParameters,
    };
    const command = new InitiateAuthCommand(input);

    const result = await client.send(command);
   
    return handleResponse({ result });
  } catch (error) {
    return handleError(error, context);
  }
};
```

LOG -> DEBUG >> body >> { email: 'TOTO123456', password: 'Azlsd45$q#s' }

### SDK version used

"@aws-sdk/client-cognito-identity-provider": "^3.245.0"

### Environment details (OS name and version, etc.)

AWS Serverless lambda

**Comment by aBurmeseDev** [comment]
Hi @smichel-amiltone - thanks for reaching out.

This error is thrown when a user isn't authorized according to service API [docs](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html#API_InitiateAuth_Errors). Can you verify that your user is authorized and credentials are correct? You can also check by running `InitiateAuth` command:
```
aws cognito-idp initiate-auth --client-id 12ab34cd56ef78gh91ij23kl45md --auth-flow USER_PASSWORD_AUTH --auth-parameters USERNAME="myuser",PASSWORD="password"
```

**Comment by smichel-amiltone** [comment]
> Hi @smichel-amiltone - thanks for reaching out.
> 
> This error is thrown when a user isn't authorized according to service API [docs](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html#API_InitiateAuth_Errors). Can you verify that your user is authorized and credentials are correct? You can also check by running `InitiateAuth` command:
> 
> ```
> aws cognito-idp initiate-auth --client-id 12ab34cd56ef78gh91ij23kl45md --auth-flow USER_PASSWORD_AUTH --auth-parameters USERNAME="myuser",PASSWORD="password"
> ```

The CMD `aws cognito-idp admin-set-user-password` accept both password

On the same user the `InitiateAuthCommand` JS command work when i have set this password Azlsd45$ and not with Azlsd45$q#s.

**Resume on the same working user !**

User required attributes : 
User name -> CL00000005 
email -> smichel@amiltone.com (Verified)

-----------------------------------------------
**Set password Azlsd45$**

```
aws cognito-idp admin-set-user-password \
--user-pool-id XXXXXXXXX \
--username CL00000005 \
--password Azlsd45$ \
--permanent \
--profil XXXXXXX  \
--region eu-central-1
```

**Run Lambda code** 

LOG -> DEBUG >> body >> { email: 'smichel@amiltone.com', password: 'Azlsd45$' }
Result InitiateAuthCommand -> Succes and `result` have authenticationResult with the token etc...

-----------------------------------------------
**Set password Azlsd45$q#s**

```
aws cognito-idp admin-set-user-password \
--user-pool-id XXXXXXXXX \
--username CL00000005 \
--password Azlsd45$q#s \
--permanent \
--profil XXXXXXX  \
--region eu-central-1
```

**Run Lambda code** 

LOG -> DEBUG >> body >> { email: 'smichel@amiltone.com', password: 'Azlsd45$q#s' }
Result InitiateAuthCommand -> NotAuthorizedException: Incorrect username or password.

**Comment by aBurmeseDev** [comment]
Thanks for the response, I'm going to reach out to service team to get their insight as this's related to service API rather than SDK. Will update once I hear back.

**Comment by darwinagyei** [comment]
Hi @aBurmeseDev - This bug is impacting my app as well. Any updates from the service team ?

**Comment by aBurmeseDev** [comment]
Service team mentioned that
>In terminal, alphanumeric characters following a dollar sign (e.g. `$abc123`) are treated as a variable name in the terminal session. Terminal will try to resolve this to a value, even if it is in the middle of a string, like the customer's `Azlsd45$q#s`. Terminal is trying to substitute the `$q` in the middle of the string with the value that is assigned to `q`.

They suspect that this might be the case:
1. Customer alls `admin-set-user-password` on CLI with a password of `Azlsd45$q#s`
2. Terminal parses the command and finds $q and assumes this is a variable
3. Terminal replaces `$q` with the value it was assigned in the terminal session
    3a. This is likely to be nothing. `$q` would be replaced with an empty string. The resulting password is then `Azlsd45#s`
4. Terminal sends the malformed password with the replaced "variable" to the AWS CLI
5. Cognito updates the user with the malformed password
6. Attempts to authenticate with any SDK and fails.

Solution is to wrap the password in the `admin-set-user-password` call in CLI with single quotes, like `'Azlsd45$q#s'`. This ensures that the `$q` does not get replaced by Terminal.

Let me know if that helps!

**Comment by smichel-amiltone** [comment]
> Service team mentioned that
> 
> > In terminal, alphanumeric characters following a dollar sign (e.g. `$abc123`) are treated as a variable name in the terminal session. Terminal will try to resolve this to a value, even if it is in the middle of a string, like the customer's `Azlsd45$q#s`. Terminal is trying to substitute the `$q` in the middle of the string with the value that is assigned to `q`.
> 
> They suspect that this might be the case:
> 
>     1. Customer alls `admin-set-user-password` on CLI with a password of `Azlsd45$q#s`
> 
>     2. Terminal parses the command and finds $q and assumes this is a variable
> 
>     3. Terminal replaces `$q` with the value it was assigned in the terminal session
>        3a. This is likely to be nothing. `$q` would be replaced with an empty string. The resulting password is then `Azlsd45#s`
> 
>     4. Terminal sends the malformed password with the replaced "variable" to the AWS CLI
> 
>     5. Cognito updates the user with the malformed password
> 
>     6. Attempts to authenticate with any SDK and fails.
> 
> 
> Solution is to wrap the password in the `admin-set-user-password` call in CLI with single quotes, like `'Azlsd45$q#s'`. This ensures that the `$q` does not get replaced by Terminal.
> 
> Let me know if that helps!

Thx for the answer, good to know

**Comment by darwinagyei** [comment]
I will implement the solution into my app. Thanks for the update!

**Comment by aBurmeseDev** [comment]
Since we've shared the response from service team and there are no other questions, I'm going to close this issue. Feel free to reach out again.

**Comment by taylor-aparai** [comment]
For anybody who comes here with similar issues, just thought I'd share the results of the last two hours of frustration in my case. 

I had the wrong client secret configured. My fault, of course. But the frustration really came from the fact that I didn't consider (wasn't aware) that Cognito implements a sort of exponential backoff rate limiting for failed login attempts. So, while I was going in circles, repeatedly attempting to log in with known good credentials, I was extending the lockout period for the user I was testing. 

Then when I noticed and fixed the client secret issue the user STILL couldn't log in and the error message was still the same. Then some period of time later the credentials "inexplicably" started working. 

Had to sit back and scratch my beard for a minute because nothing in the Cognito console indicated that the user was locked out for failed attempts. 

Found this in the docs: 

> Amazon Cognito implements a temporary account lockout mechanism after multiple failed sign-in attempts. After five failed attempts, a user is locked out for one second. The lockout duration doubles with each subsequent failed attempt, with a maximum lockout of about 15 minutes. Additional attempts during a lockout won't change the duration. The lockout is cleared by a successful sign-in after the lockout expires or by not attempting to sign in for 15 minutes. This behavior may change.
